public/app/plugins/datasource/prometheus/querycache/QueryCache.ts

```diff
@@ -62,26 +62,60 @@ function getTargSig(targExpr: string, request: DataQueryRequest<PromQuery>, targ
  * Ident: Identity: the string that is not expected to change
  * Sig: Signature: the string that is expected to change, upon which we wipe the cache fields
  */
-type Milliseconds = number;
 export class QueryCache {
-  private overlapWindow: Milliseconds;
+  private overlapWindowMs;
+  private perfObeserver: PerformanceObserver;
+  private pendingRequestIds = new Set<string>();
+
+  cache = new Map<TargetIdent, TargetCache>();
+
   constructor(overlapString?: string) {
     const unverifiedOverlap = overlapString ?? defaultPrometheusQueryOverlapWindow;
+
     if (isValidDuration(unverifiedOverlap)) {
       const duration = parseDuration(unverifiedOverlap);
-      this.overlapWindow = durationToMilliseconds(duration);
+      this.overlapWindowMs = durationToMilliseconds(duration);
     } else {
       const duration = parseDuration(defaultPrometheusQueryOverlapWindow);
-      this.overlapWindow = durationToMilliseconds(duration);
+      this.overlapWindowMs = durationToMilliseconds(duration);
     }
-  }
 
-  cache = new Map<TargetIdent, TargetCache>();
+    this.perfObeserver = new PerformanceObserver((list) => {
+      list.getEntries().forEach((entry) => {
+        if (entry.initiatorType === 'fetch') {
+          let fetchUrl = entry.name;
+
+          if (fetchUrl.includes('/api/ds/query')) {
+            let match = fetchUrl.match(/requestId=([a-z\d]+)/i);
+
+            if (match) {
+              let requestId = match[1];
+
+              if (this.pendingRequestIds.has(requestId)) {
+                // TODO: store full initial request size by targSig so we can diff follow-up partial requests
+                // TODO: log savings between full initial request and incremental request to Faro
```

Comment:
this seems like somethign that can be done in post processing on loki IIUC

Reply:
`requestId`s are not currently unique, they're just sequential from initial dashboard load, e.g. `Q100`, `Q101`. then `Q103` might be initiated by another panel/DS that doesnt need tracking (e.g. Influx), so it will be untagged and omitted from this Prom measuring path.

if we actually had request UIDs, then yeah you can maybe do something in loki. but right now you'll just have a bunch of `Q100`-tagged requests that are unrelated.